src/jupyterlabcontroller/services/lab.py

```diff
@@ -0,0 +1,204 @@
+import base64
+from collections import deque
+from dataclasses import dataclass
+from typing import Dict, List
+
+from aiojobs import Scheduler
+
+from ..models.v1.domain.config import LabSecrets
+from ..models.v1.domain.context import ContextContainer, RequestContext
+from ..models.v1.external.lab import LabSpecification, UserQuota
+from ..storage.k8s import NetworkPolicySpec, NSCreationError, PodSpec, Secret
+from ..utils import get_namespace_prefix
+
+
+@dataclass
+class LabManager:
+    lab: LabSpecification
+    nublado: ContextContainer
+    context: RequestContext
+
+    async def check_for_user(self) -> bool:
+        """True if there's a lab for the user, otherwise false."""
+        return self.context.user.username in self.nublado.user_map
+
+    async def create_lab(self) -> None:
+        """Schedules creation of user lab objects/resources."""
+        username = self.context.user.username
```

Comment:
I find this confusing.  Isn't the whole point of having a request context dependency exactly to encapsulate the stuff that comes in on a given request?

I already have a reasonable method for faking a request context object for doing unit testing of my various methods.  It seems to me like breaking apart the context dependency at the handler layer and passing the pieces individually to the various service methods just complicates both the handlers and the service methods.

Reply:
The problem is that you're leaking a ton of data across the API boundary, so the components are not staying as black boxes with minimum inputs and therefore complexity is leaking through the code base. The goal with a large application like this is to try to keep the components as black boxes defined only by their APIs and define those APIs to be as minimal as possible so that there are fewer variables and less state you have to keep in memory when working on a part of the code.

In particular here, there are two problems: one is that you're injecting the entirety of `RequestContext` into the service, so now the service may use any part of that object and anything you add to that object later, rather than only having access to parameters you pass in and dependencies that are injected. The second problem is that methods on service objects should be sentences where the method is a verb and the parameters are the objects the verb acts on."Get this" or "create this" or "delete this." But here, the objects are just missing; from the perspective of the caller or someone editing this code, the objects come from the ether, from this context that was injected earlier from who knows where (because the context is lost at the point of the method call or method implementation).

It looks like more complexity, but I think it really is less complexity if the call says "create a lab for this user" as opposed to "create a lab (using unmentioned nebulous semi-global state)". It is somewhat more repetition, but what you get for that repetition is clarity when reading the code in knowing that the method is only going to act based on the stuff that you pass into the method when you call it.

Reply:
It turned out to be a larger task than I anticipated, but I have diced up the dependencies and reworked injection.